---
title: "Deconvolution using Scaden"
author: "Karishma D'Sa"
output: 
  html_document:
    code_folding: hide
    df_print: paged
    toc: yes
    toc_float: yes
    toc_depth: 6
  html_notebook:
    code_folding: hide
    toc: yes
    toc_float: yes
---

#Scaden overview
Scaden (a deep neural network) uses gene expression information to infer cellular composition of tissues. <br>
Menden et. al, Science Advances 2020, https://www.science.org/doi/10.1126/sciadv.aba2619
<br><br>
Scaden uses simulated bulk RNA_Seq samples, generated from the single cell data,for training and predicts cell type proportions.  <br>


![Scaden overview](/home/kdsa/ipscAstrocyteNeuron/files/scaden_overview.jpeg)

Overview of training data generation and cell type deconvolution with Scaden.(A) Artificial bulk samples are generated by subsampling random cells from an scRNA-seq dataset and merging their expression profiles. (B) Model training and parameter optimization on simulated tissue RNA-seq data by comparing cell fraction predictions to ground-truth cell composition. (C) Cell deconvolution of real tissue RNA-seq data using Scaden.
<br><br>
Scaden's workflow ( described in https://scaden.readthedocs.io/en/latest/usage.html):<br>

1.Generate training data (artificial bulk samples) from the scRNA-Seq datasets. <br>
 
 <u>scRNA-seq data processing</u>: 
<i>Counts file </i>: Normalize count data to library size, count data should be of the size nSamples x nGenes, <i>cell type labels file </i> with a single column labelled 'Celltype' the order must be the same as that in the counts file. <br><br>
Run <i>scaden simulate </i> to genrate the artifical bulk samples for training. <br>

2. process : scaden performs pre-processing on the training data. A new file for training which only contains the intersection of genes between the training and the prediction data. #Also the training data will be log2-transformed and scaled to the range [0,1].

Options:
  --var_cutoff FLOAT     Filter out genes with a variance less than the
                         specified cutoff. A low cutoff is recommended,this
                         should only remove genes that are obviously
                         uninformative.



3. train : Training a Scaden ensemble model. <br>
It is trained on data containing both the gene expression and real cell type fraction details. Scaden  is an ensemble of the 3 best performing models and the final cell type composition estimates are the averaged predictions of the 3 models. <br>


<br>
Scaden consists of three deep neural network models. By default, each model is trained for 5,000 steps ( recommended number by authors- used this for datasets of ~ 30K samples) <br>
<br><br>
The process summary - <i> <br>
"- Each of the three ensemble models is trained independently for 5000 steps. <br>
 -This “early stopping” serves to avoid domain overfitting on the simulated tissue data, which would decrease the model performance on the real tissue RNA-seq data.<br>
 -To perform deconvolution with Scaden, a bulk RNA-seq sample is fed into a trained Scaden ensemble, and three independent predictions for the cell type fractions of this sample are generated by the trained DNNs. <br>
 -These three predictions are then averaged per cell type to yield the final cell type composition for the input bulk RNA-seq sample"
</i>

<br>
4. predict: The prediction file can contain either raw or normalized counts. They should not be in logarithmic space. 


```{r libraries, eval = T, include = F}
library(here)
library(DESeq2)
library(tidyverse)
library(knitr)
library(gridExtra)
file.path(here())
source(here("R","common_dge.R"))
source(here("R","common_deconv.R"))
source(here("R/plot_setting.R"))
```
#Implementing Scaden
##Pre-processing data
###scRNA-Seq data
<br>
File format for the sc data required: <br>
(a) Cell type labels file : should be of size (n x 1), where n is the number of cells in the data.<br>
The single column in this file has to be labeled ‘Celltype’. Extra columns may be present, as long as the ‘Celltype’ column which specifies the cell type labels in the correct order is present.<br>

(b)Counts file : count data should be of size (n x g), where g is the number of genes and n is the number of samples. The order must be the same as for the cell type labels.
<br><br>

Scaden requires the normalised or raw counts of the scRNA-Seq data (https://scaden.readthedocs.io/en/latest/usage.html#file-formats).
<br><br>
<b>Note:</b> Generating a single file containgin all the cells from the 6 samples with the raw counts for scaden.(sc_R_analysis/2-2.genData_4Scaden.Rmd) 

<br>
Clustering resolution 0.25 was used.
<br><br>


###Pre-processing bulk tissue data
This should be a file of shape m X n, where m are your features (genes) and n your samples. So each row corresponds to a gene, and each column to a sample. The column name for the genes should be empty. <br><br>

Data can either be raw counts or normalized ( scaden can handle both), just make sure that they are not in logarithmic space already as Scaden applies its scaling procedure to it, which involves taking the logarithm of the counts. <br>
<br>
<br>
Generating the bulk counts file <br>
1. As the single cell data file has the gene symbols, excluding ENSG ids that no not have a gene symbol <br>
2. Also in case of genes with same symbol having different ENSG ids, taking the ENSG where the average normalised counts is the highest.
3. Generating a raw counts, normalised and scaled normalised counts files.

```{r pre-proc bulk tissue data, eval=F, include = F}
load(here("../results/dge/deseq2DataSet_filtered_collapsedReplicates.rda"))
dds <- ddsCollapsed
rm(ddsCollapsed)

#get the library size
lib_matrix <- colSums(counts(dds)) %>% 
    as.matrix() %>% 
    t()

#norm counts scale = counts/library size * 1000000 with scaling 
scaled_nml_counts <- as.data.frame( counts(dds)/c(lib_matrix) * 1000000)
#raw counts
raw_counts <- as.data.frame(counts(dds))

#Filtering genes
###############
#get the gene symbols
geneInfo <- getGeneNameBiotype(rownames(raw_counts), chrName=T) #return a data table
geneInfo <- geneInfo[ chromosome_name %in% c(1:22, "X", "Y", "MT"), .(ensembl_gene_id, external_gene_name) ] 
# Filter out rows where ensembl ID has no corresponding gene symbol
geneInfo <- geneInfo[external_gene_name != "", ]

#take the record with highest average normalised counts where duplicate hgnc_symbols (i.e. multiple ensembl ids mapping to same hgnc_symbol)
scaled_nml_counts$geneid <- rownames(scaled_nml_counts)
setDT(scaled_nml_counts)
scaled_nml_counts[,mean:=rowMeans(.SD), ,by=geneid]

#get the row with the max counts
gene_ct <- scaled_nml_counts[ , .(geneid, mean)]
setDT(geneInfo)[gene_ct, mean := mean, on = c(ensembl_gene_id="geneid") ] #join

gene_max <- geneInfo[geneInfo[, .I[mean == max(mean)], by=external_gene_name]$V1] #max value row

#saving the counts for genes that pass the filter,  adding the gene symbols
##################
setDF(geneInfo)
geneInfo$mean <- NULL

raw_counts <- raw_counts[ rownames(raw_counts)%in% gene_max$ensembl_gene_id,]
raw_counts <- raw_counts %>%
          rownames_to_column(var= "ensembl_gene_id") %>%
  left_join(geneInfo, by = "ensembl_gene_id") %>%
  column_to_rownames(var ="external_gene_name") %>%
  select(-c("ensembl_gene_id"))

scaled_nml_counts <- as.data.frame( counts(dds)/c(lib_matrix) * 1000000)
scaled_nml_counts <- scaled_nml_counts[ rownames(scaled_nml_counts)%in% gene_max$ensembl_gene_id,]
scaled_nml_counts <- scaled_nml_counts %>%
          rownames_to_column(var= "ensembl_gene_id") %>%
  left_join(geneInfo, by = "ensembl_gene_id") %>%
  column_to_rownames(var ="external_gene_name") %>%
  select(-c("ensembl_gene_id"))

#saving counts
bulk_counts_path<- here("../results/deconv_scaden/input/bulk_data/") 
head( format(nml_counts, scientific = FALSE),1)
write.table(raw_counts ,
            file = paste0(bulk_counts_path, "ipsc_bulk_raw.txt"), 
            sep = "\t", row.names = T, 
            quote=F)
#scaled
write.table(format(scaled_nml_counts, scientific=FALSE) ,
            file = paste0(bulk_counts_path, "ipsc_bulk_scaled_nml.txt"), 
            sep = "\t", row.names = T, 
            quote=F)
```

```{r no_of_genes, eval=T, echo=F }
bulk_count_file <- here("../results/deconv_scaden/input/bulk_data/ipsc_bulk_raw.txt")
dt <- read.table(bulk_count_file)
cat("\n Number of genes = ",nrow(dt)) 
```

# scRNASeq samples
Providing Scaden with a single file containing all the cells ( all cultures) with raw counts and celltypes based on the clustering 0.25- <br> <br>
This also allows an increase in the nCells that can be used to create the training datasets <br>

Rawcounts of all samples from the clustering resolution 0.25 combined in a single file <br>
8132 cells ; 22588 genes <br>

## Raw counts
<b> Varying nCells </b>
Running Scaden increasing the no. of cells <br> 
nCells = (100 500 1000 2000 3000 4000 4500 5000) <br>
script : run_scaden_deconv.sh <br> <br>
<br>
21242 common genes at variance cutoff 0.1

```{bash check common genes, eval = F, include = F }
grep "common genes" ~/ipscAstrocyteNeuron/ipscAstroNeu/logs/scaden_opt_bulkRaw_scCombRaw*.log
 
```
```{r combSamples raw sc and raw bulk counts_ plot varying nCells var_cutoff, eval=F, include=F }
pred_file_prefix <- "scaden_pred_comb_"
pred_path <- here("../results/deconv_scaden/allsamples_run/predictions/")
scaden_celltype_prop <- get_scaden_predictions(pred_path, pred_file_prefix)

sc_celltype_prop <- get_sc_celltype_prop_with_indi_info(sc_celltypeprop_path = here("../results/deconv_scaden/input/") )
  
comb_dt <- rbindlist(list(sc_celltype_prop, scaden_celltype_prop), use.names = T)

mdata <- melt(comb_dt, id=c(names(comb_dt)[ grep("celltype_", names(comb_dt), invert = T)] ))
setnames(mdata, old = c("variable", "value"), new = c("cell_type", "cell_type_prop"))
mdata$cell_type <- (gsub("celltype_", "", mdata$cell_type))  


mdata$colour <- rep('black', nrow(mdata))
#mdata$colour[mdata$cells == "neuron"] <- "neuron"
mdata$colour[mdata$indi == "ctrl_1_JOM_inhouse"] <- "blue"
mdata$colour[mdata$indi == "ctrl_NTN"] <- "green"

head(mdata, 3)
col_vals <- as.character(mdata$colour)
names(col_vals) <- as.character(mdata$indi)
source(here("./R/plot_setting.R"))
 
mdata$nCells <- as.numeric(mdata$nCells)

 ggplot( mdata, aes(x = cell_type, y = cell_type_prop*100), col= indi, group = cell_type ) +
     geom_boxplot(aes( fill= cell_type) ) +
   geom_point(aes(col = indi), position = position_jitterdodge(jitter.width = 0.05, jitter.height = 0.05), shape=1 )+
   scale_color_manual(values = col_vals) +
  facet_grid(cells~nCells) +
   ggtitle("Scaden's cell type proportion Varying nCells using raw counts - bulk and sc data \n NA = cell type proportions from the scRNAseq\n single sc data file used to create training data simulating the bulk")
   
  
ggsave(paste0(here("../results/deconv_scaden/allsamples_run/plots/"), "celltype_prop_var01_varyncells_scdata_comb_scIndi.png"), dpi = 300, height = 8, width = 10)
  
ggplot( mdata, aes(x = cell_type, y = cell_type_prop*100,  col = cell_type) ) +
     geom_boxplot(aes( fill= cell_type) ) +
  facet_grid(cells~nCells) +
   ggtitle("Scaden's cell type proportion Varying nCells using raw counts - bulk and sc data \n NA = cell type proportions from the scRNAseq\n single sc data file used to create training data simulating the bulk")
   
  
ggsave(paste0(here("../results/deconv_scaden/allsamples_run/plots/"), "celltype_prop_var01_varyncells_scdata_comb.png"), dpi = 300, height = 8, width = 10)

rm(comb_dt)


```


```{r display comb sc raw counts and raw bulk plots varying cut off, eval=T, echo=F, out.width="70%", out.height="70%" }
opt_plot_files <- here("../results/deconv_scaden/allsamples_run/plots/celltype_prop_var01_varyncells_scdata_comb.png")
include_graphics(opt_plot_files)
```

```{r displayplots scclust _celltype_prop, eval=T, echo=F }
files <- here("../results/deconv_scaden/input/sc_data/umap_clust.png")
knitr::include_graphics(files)
```


From the plots above can see in the case of ncells >= 3000 the proportions seem to match better .<br>
<br>

<b> Varying no. of bulk simulated samples </b> as in the above runs using a single sc data file, we only create 1000 bulk simulated samples.
<br> Using nCells =3000 as seen in the tables below, the correlation of the cell type proportion from the sc data and scaden is high. <br>
nSamples = (2000 3000 5000) <br>
```{r display comb sc raw counts and bulk plots varying nSamples, eval=T, echo=F, out.width="70%", out.height="70%" }
opt_plot_files <- list.files( here("../results/deconv_scaden/allsamples_run/plots/"), pattern="*_varynsamples_scdata_comb.png", full.names = T)

include_graphics(opt_plot_files)
```

Correlation table [Matched controls, varying nSamples, variance 0.1, nCells = 3000](#raw-vary-nsamples)
<br>


### Correlation of the cell type proportions
As there is a matching control, viewing the correlation of the celltype proportions as per scRNAseq and scaden's deconvolution for the bulk for those samples <br>

"ctrl_1_JOM_inhouse", "ctrl_NTN" <br>

From the plots below can see in the case of ncells > 3000 the proportions seem to match better except in the co-culture.<br>

#### Varying nCells

```{r plot_celltype prop in sc vs bulk for matched controls, eval =F, echo =F}
ctrls2sel <- c("ctrl_1_JOM_inhouse", "ctrl_NTN")
pred_file_prefix <- "scaden_pred_comb_"

pred_path <- here("../results/deconv_scaden/allsamples_run/predictions/")
var2sel <- 0.1

sc_prop_path <- here("../results/deconv_scaden/input/")

title <- paste0("For matched individuals/controls - Celltype proportions in the scRNA_seq data vs that called by scaden \n at var_cutoff ", var2sel, "and varying nCells \n Raw counts")
 outPng <- here("../results/deconv_scaden/allsamples_run/plots/matchedCtrls_scVsScad_celltype_prop_var01_varyncells_scdata_comb_new.png")
 
ret <- generate_plot_ctProp_scVsScad ( pred_path, pred_file_prefix, sc_prop_path, var2sel ,  title, outPng, matchedCtrls=ctrls2sel)
   
```


```{r displayplot ctprop sc vs sscade matched ctrl rawcounts, eval=T, echo=F}
files <- here("../results/deconv_scaden/allsamples_run/plots/matchedCtrls_scVsScad_celltype_prop_var01_varyncells_scdata_comb.png")
knitr::include_graphics(files)
```

```{r correlation of cell type prop with sc data, eval = F, echo=F}
# checking the correlation of the cell type prop
pred_file_prefix <- "scaden_pred_comb_"
pred_path <- here("../results/deconv_scaden/allsamples_run/predictions/")
scaden_celltype_prop <- get_scaden_predictions(pred_path, pred_file_prefix)

sc_celltype_prop <- get_sc_celltype_prop_with_indi_info(sc_celltypeprop_path = here("../results/deconv_scaden/input/") )

scaden_celltype_prop[ , cell_treat := paste0(cells, "_", treatment)]
sc_celltype_prop[ , cell_treat := paste0(cells, "_", treatment)]
corr_sc_bulk <- get_corr_sc_scaden ( sc_celltype_prop, scaden_celltype_prop)

outFile <- here("../results/deconv_scaden/allsamples_run/spearman_corr_sc_scaden_celltype_prop.tsv")
fwrite(corr_sc_bulk, outFile, sep="\t", row.names = F, quote=F)

```
#### Varying nSamples, nCells = 3000

```{r plot_celltype prop in sc vs bulk for matched controls varying nSamples, eval =F, echo =F}
ctrls2sel <- c("ctrl_1_JOM_inhouse", "ctrl_NTN")
pred_file_prefix <- "scaden_pred_comb_"

pred_path <- here("../results/deconv_scaden/allsamples_run/predictions_nsamples/")
var2sel <- 0.1
nCells2sel <- 3000

sc_prop_path <- here("../results/deconv_scaden/input/")

title <- paste0("For matched individuals/controls - Celltype proportions in the scRNA_seq data vs that called by scaden \n at var_cutoff ", var2sel, ", nCells =", nCells2sel, " and varying nSamples \n Raw counts")
 outPng <- here("../results/deconv_scaden/allsamples_run/plots/matchedCtrls_scVsScad_celltype_prop_var01_3000cells_varynSamples_scdata_comb.png")
 
ret <- generate_plot_ctProp_scVsScad ( pred_path, pred_file_prefix, sc_prop_path, var2sel ,  title, outPng, matchedCtrls=ctrls2sel, nCells2sel, extract_nSamples_from_fname =T, facetbyCells =F)
   
```

```{r displayplot vary nsamples ctprop sc vs sscade matched ctrl rawcounts, eval=T, echo=F}
files <- here("../results/deconv_scaden/allsamples_run/plots/matchedCtrls_scVsScad_celltype_prop_var01_3000cells_varynSamples_scdata_comb.png")
knitr::include_graphics(files)
```

### Tabular view of the Spearman rank-order correlation 
<br>
Checking the correlation of the cell type proportions in the scRNAseq data with each sample's deconvolution by scaden. (Spearman correlation(two-sided)) <br>
Viewing the samples common to the scRNAseq and bulk - expect it to be highest

#### Matched samples, varying nCells - Scaden var_cutoff = 0.1 {.tabset .tabset-fade .tabset-pills}

From the tables below can see the correlation is high for all samples, except neuron untreated, at nCells =5000
```{r view correlations matched samples, eval=T, include = F}
corrFile <- here("../results/deconv_scaden/allsamples_run/spearman_corr_sc_scaden_celltype_prop.tsv")

var_sel <- 0.1
nCells_sel <- c(100, 3000, 4000, 5000)
corr_dt <- fread(corrFile)
corr_dt <- corr_dt[ indi %in% c("ctrl_1_JOM_inhouse", "ctrl_NTN") & (ncells %in% nCells_sel) & (var_cutoff %in% var_sel), ]
#head(corr_dt, 3)
corr_dt$ncells <- as.character(corr_dt$ncells)
data_source <- unique(corr_dt$ncells)
out <- vector(mode = "character", length = length(data_source))
for (i in data_source) {
  out[i] <- knit_expand(text = c("##### {{stringr::str_to_title(i)}}",
                                 "```{r, echo = FALSE}",
                                 "corr_dt[ncells == '{{i}}',.(indi, cell_treat, corr, corrPval)]",
                                 "```"))
}

```
`r paste(knit(text = out), collapse = '\n')`

#### Looking at the correlations in all samples - scaden run at var_cutoff 0.1 and nCells 3000 {.tabset .tabset-fade .tabset-pills}

Except for the neuronal untreated sample, there is a high  correlation in the cell type proportions from scRNAseq data and deconvolution by scaden. <br>
Can see the correlation in the case of the commercial controls differs though not to a large extent from the inhouse controls (astrocyte treated and the coculture samples)

```{r view correlations nCells 3000, eval=T, include = F}
corrFile <- here("../results/deconv_scaden/allsamples_run/spearman_corr_sc_scaden_celltype_prop.tsv")

var_sel <- 0.1
nCells_sel <- c(3000)
corr_dt <- fread(corrFile)
corr_dt <- corr_dt[ (ncells %in% nCells_sel) & (var_cutoff %in% var_sel), ]
data_source <- unique(corr_dt$cell_treat)
out <- vector(mode = "character", length = length(data_source))
for (i in data_source) {
  out[i] <- knit_expand(text = c("##### {{stringr::str_to_title(i)}}",
                                 "```{r, echo = FALSE}",
                                 "corr_dt[cell_treat == '{{i}}',.(indi, cell_treat, corr, corrPval)]",
                                 "```"))
}

```
`r paste(knit(text = out), collapse = '\n')`

#### Matched controls, varying variance cutoff 0.2, 0.3 and nCells =3000 {.tabset .tabset-fade .tabset-pills}
At 0.2 , 21242 common genes <br>
at 0.3, 21241 <br>
Not much difference from <br>
21242 common genes at variance cutoff 0.1 <br>
<br>

```{r matched sample correlations var 0.1 0.2 0.3 ncells 3000, eval=T, echo =F}
pred_file_prefix <- "scaden_pred_comb_"
pred_path <- here("../results/deconv_scaden/allsamples_run/predictions/")
scaden_celltype_prop <- get_scaden_predictions(pred_path, pred_file_prefix, extract_nSamples_from_fname = T)
scaden_celltype_prop <- scaden_celltype_prop[ nCells == 3000, ]
sc_celltype_prop <- get_sc_celltype_prop_with_indi_info(sc_celltypeprop_path = here("../results/deconv_scaden/input/") )

scaden_celltype_prop[ , cell_treat := paste0(cells, "_", treatment)]
sc_celltype_prop[ , cell_treat := paste0(cells, "_", treatment)]
corr_sc_bulk <- get_corr_sc_scaden ( sc_celltype_prop, scaden_celltype_prop)

corr_dt <- corr_sc_bulk[ indi %in% c("ctrl_1_JOM_inhouse", "ctrl_NTN") , ]

corr_dt$var_cutoff <- as.character(corr_dt$var_cutoff)
#view the correlations
data_source <- unique(corr_dt$var_cutoff)
out <- vector(mode = "character", length = length(data_source))
for (i in data_source) {
  out[i] <- knit_expand(text = c("##### {{stringr::str_to_title(i)}}",
                                 "```{r, echo = FALSE}",
                                 "corr_dt[var_cutoff == '{{i}}',.(indi, cell_treat, ncells, corr, corrPval)]",
                                 "```"))
}

```
`r paste(knit(text = out), collapse = '\n')`

#### <a id="raw-vary-nsamples"></a> Matched controls, varying nSamples, variance 0.1, nCells = 3000 {.tabset .tabset-fade .tabset-pills}
Varying no. of bulk simulated samples as in the above runs using a single sc data file, we only create 1000 bulk simulated samples.
<br> Using nCells =3000  <br>
nSamples = (2000 3000 5000) <br>

Increasing to include 8000, 10000 <br>

```{r combSamples raw sc and bulk counts_ plot varying nSamples , eval=F, include=F }
pred_file_prefix <- "scaden_pred_comb_"
pred_path <- here("../results/deconv_scaden/allsamples_run/predictions_nsamples/")
scaden_celltype_prop <- get_scaden_predictions(pred_path, pred_file_prefix, extract_nSamples_from_fname = T)
head(scaden_celltype_prop, 2)
run_ncells <- 3000 #unique(scaden_celltype_prop$nCells)
var_thresh <- 0.1

#get the prediciton for 1000 nSamples for completion
scaden_1000 <- get_scaden_predictions(here("../results/deconv_scaden/allsamples_run/predictions/"), pred_file_prefix, extract_nSamples_from_fname = F)
scaden_1000 <- scaden_1000[ var_cutoff == var_thresh & nCells == run_ncells, ]
scaden_1000$nSamples <- "1000"  

sc_celltype_prop <- get_sc_celltype_prop_with_indi_info(sc_celltypeprop_path = here("../results/deconv_scaden/input/") )
sc_celltype_prop$nSamples <- "scRNA-seq"  
comb_dt <- rbindlist(list(sc_celltype_prop, scaden_1000, scaden_celltype_prop), use.names = T)

mdata <- melt(comb_dt, id=c(names(comb_dt)[ grep("celltype_", names(comb_dt), invert = T)] ))
setnames(mdata, old = c("variable", "value"), new = c("cell_type", "cell_type_prop"))
mdata$cell_type <- (gsub("celltype_", "", mdata$cell_type))  


mdata$colour <- rep('black', nrow(mdata))
#mdata$colour[mdata$cells == "neuron"] <- "neuron"
mdata$colour[mdata$indi == "ctrl_1_JOM_inhouse"] <- "blue"
mdata$colour[mdata$indi == "ctrl_NTN"] <- "green"

head(mdata, 3)
col_vals <- as.character(mdata$colour)
names(col_vals) <- as.character(mdata$indi)
source(here("./R/plot_setting.R"))
 
mdata$nSamples <- as.numeric(mdata$nSamples)

title <- paste0("Scaden's cell type proportion varying nSamples using raw counts - bulk and sc data \n NA = cell type proportions from the scRNAseq\n single sc data file used to create training data simulating the bulk \n var =0.1; ncells =", run_ncells)

 ggplot( mdata, aes(x = cell_type, y = cell_type_prop*100), col= indi, group = cell_type ) +
     geom_boxplot(aes( fill= cell_type) ) +
   geom_point(aes(col = indi), position = position_jitterdodge(jitter.width = 0.05, jitter.height = 0.05), shape=1 )+
   scale_color_manual(values = col_vals) +
  facet_grid(cells~nSamples) +
   ggtitle(title)
   
  
ggsave(paste0(here("../results/deconv_scaden/allsamples_run/plots/"), "celltype_prop_var01_cells", run_ncells, "_varynsamples_scdata_comb_scIndi.png"), dpi = 300, height = 8, width = 10)
  
ggplot( mdata, aes(x = cell_type, y = cell_type_prop*100,  col = cell_type) ) +
     geom_boxplot(aes( fill= cell_type) ) +
  facet_grid(cells~nSamples) +
   ggtitle(title)
   
  
ggsave(paste0(here("../results/deconv_scaden/allsamples_run/plots/"), "celltype_prop_var01_cells", run_ncells, "_varynsamples_scdata_comb.png"), dpi = 300, height = 8, width = 10)

rm(comb_dt)
```


```{r spearman correlations varying nsamples ncells 3000, eval=T, echo =F}
pred_file_prefix <- "scaden_pred_comb_"
pred_path <- here("../results/deconv_scaden/allsamples_run/predictions_nsamples/")
scaden_celltype_prop <- get_scaden_predictions(pred_path, pred_file_prefix, extract_nSamples_from_fname = T)
scaden_celltype_prop <- scaden_celltype_prop[var_cutoff == 0.1 & nCells == 3000, ]


#get the prediciton for 1000 nSamples for completion
scaden_1000 <- get_scaden_predictions(here("../results/deconv_scaden/allsamples_run/predictions/"), pred_file_prefix, extract_nSamples_from_fname = F)
scaden_1000 <- scaden_1000[ var_cutoff == 0.1 & nCells == 3000, ]
scaden_1000$nSamples <- "1000"  
scaden_celltype_prop <- rbindlist(list(scaden_1000,scaden_celltype_prop), use.names = T)

scaden_celltype_prop$nCells <- scaden_celltype_prop$nSamples
sc_celltype_prop <- get_sc_celltype_prop_with_indi_info(sc_celltypeprop_path = here("../results/deconv_scaden/input/") )

scaden_celltype_prop[ , cell_treat := paste0(cells, "_", treatment)]
sc_celltype_prop[ , cell_treat := paste0(cells, "_", treatment)]
corr_sc_bulk <- get_corr_sc_scaden ( sc_dt = sc_celltype_prop, scaden_dt = scaden_celltype_prop)
setnames(corr_sc_bulk, old = "ncells", new = "nSamples")

outFile <- here("../results/deconv_scaden/allsamples_run/spearman_corr_sc_scaden_celltype_prop_nCells3000_varynSamples.tsv")
fwrite(corr_sc_bulk, outFile, sep="\t", row.names = F, quote=F)
head(corr_sc_bulk, 2)
dcast_corr <- dcast(corr_sc_bulk, method+indi+var_cutoff+cell_treat~nSamples, value.var = c("corr", "corrPval"))
dcast_corr$nCells <- 3000
outFile <- gsub(".tsv", "_wideform.tsv", outFile)
fwrite(dcast_corr, outFile, sep="\t", row.names = F, quote=F)

```

```{r matched sample correlations varying nsamples ncells 3000, eval=T, echo =F}

corr_dt <- fread(here("../results/deconv_scaden/allsamples_run/spearman_corr_sc_scaden_celltype_prop_nCells3000_varynSamples.tsv"))
corr_dt <- corr_dt[ indi %in% c("ctrl_1_JOM_inhouse", "ctrl_NTN") , ]

corr_dt$nSamples <- as.character(corr_dt$nSamples)
#view the correlations
data_source <-(unique(corr_dt$nSamples))
out <- vector(mode = "character", length = length(data_source))
for (i in data_source) {
  out[i] <- knit_expand(text = c("##### {{stringr::str_to_title(i)}}",
                                 "```{r, echo = FALSE}",
                                 "corr_dt[nSamples == '{{i}}',.(indi, cell_treat, corr, corrPval)]",
                                 "```"))
}

```
`r paste(knit(text = out), collapse = '\n')`
